### ISSUE TITLE
Allow custom webpack config

### ISSUE THREAD

**Issue by thomasdigby** [shortlist-digital/tapestry-wp#68]
We should discuss this as I'm fairly certain we'll need it. We'll need to decide what implementation to use, do we try and gracefully merge the user config into our default config, or do we overwrite it completely?

Can we overwrite it completely? Probably not, I assume we'll need to set some defaults, e.g. entry, output, aliases, babel loader at a minimum.

Do we add a webpack section to our `tapestry.config.js`, do we allow them to create a `webpack.config.js` in the project root (it'll be an incomplete and invalid webpack config though)? 

Do we consume an object? Do we expose a function to modify the webpack config?

Thoughts needed. 🌮

**Comment by jonsherrard** [comment]
I think a merge is too prone to error, and we need to "own" the webpack config, it's an opinionated library after all.

if projectConfig.webpack typeof function:
call `webpackConfig = projectConfig.webpack(internalTapestryWebpackConfig)`

**Comment by thomasdigby** [comment]
So my current thinking of this is such:

- We'll need to bundle both the server and the client with the same custom webpack, any `resolve.alias`ing and the like will be required by both
- That should be simple enough, we can include a webpack node API instance in the bin and bundle it before running
- However, we can't pass the custom config through `tapestry.config.js` as we might only be able to parse the tree once it's altered via webpack (resolving modules etc.). If we `const config = require('tapestry.config.js')` config might not be able to parse an import that needs resolving through webpack.
- That means we'll need to store the custom config elsewhere, `tapestry.webpack.js` or just `webpack.config.js` in the root?
- As per [this lovely snippet](https://github.com/palmerhq/backpack/blob/master/packages/backpack-core/bin/dev#L23) we'll also need to pass through some data to the custom config, like `cwd`, `env`, the webpack instance and the like.
- Which means that it'll need to be a function e.g.
```
module.exports = (config, { cwd, env }, webpack) => {
  // modify config
  return config
}
```
- However this will be majorly prone to errors unless they stick to a small set of options to modify, and they return the config, and they check that keys exist on the object before setting values

`</braindump>`

**Comment by jonsherrard** [comment]
This is a most excellent brain dump.

Separate webpack.config.js feels good.

I think by it's nature this is an `advanced` feature, Tapestry is super opinionated like a lot of these tools, and we include a fair chunk of stuff out of the box. If it's well documented and we have an example... I think we're OK.

I'd be very happy with a working solution that utilised `module.exports = { cwd, env }, webpack) =>`

Definitely needs a "Your mileage may vary" label.

**Comment by thomasdigby** [comment]
Fixed by #85
